File: files/2024-06-30/CSEC 30.06.R
library(tidyverse)
library(readxl)

path = "files/Excel Challenge  30th June.xlsx"

input = read_xlsx(path, range = "B3:D7")
test  = read_xlsx(path, range = "E3:I7")

result = input %>%
  mutate(seq = map2(`Start Date`, `End Date`, seq, by = "month")) %>%
  unnest_longer(seq) %>%
  mutate(year = year(seq),
         val = 1) %>%
  select(Project, year, val) %>%
  pivot_wider(names_from = year, values_from = val, values_fn = sum) %>%
  select(-Project)

identical(result, test)
# [1] TRUE

File: files/2024-06-30/CSEC 30.06.py
import pandas as pd

path = "files/Excel Challenge  30th June.xlsx"

input = pd.read_excel(path, usecols="B:D", skiprows=2, nrows=4)
test = pd.read_excel(path, usecols="E:I", skiprows=2, nrows=4).fillna(0).astype(int)

result = input.copy()
result['seq'] = result.apply(lambda x: pd.date_range(start=x["Start Date "], end=x["End Date"], freq='M'), axis=1)
result = result.explode('seq')
result['year'] = result['seq'].dt.year
result['val'] = 1
result = result[['Project', 'year', 'val']].\
    pivot_table(index='Project', columns='year', values='val', aggfunc='sum').\
    fillna(0).astype(int)
result = result.reset_index().drop(columns='Project')

print(result.equals(test)) # True